compiler/AST/ImportStmt.cpp

```diff
@@ -233,3 +260,143 @@ bool ImportStmt::checkValid(Expr* expr) const {
+*                                                                             *
+************************************** | *************************************/
+
+bool ImportStmt::providesQualifiedAccess() const {
+  /* Today, ImportStmts only support qualified access of all symbols in the
+     imported modules, or supports unqualified access of certain symbols in it,
+     but not both.  There are plans for enabling both if we desire it (by
+     allowing `this` in ImportStmt's unqualified access lists)
+  */
+  if (unqualified.size() == 0) {
+    return true;
+  } else {
+    return false;
+  }
+}
+
+/************************************* | **************************************
+*                                                                             *
+* Returns true if the current import statement has the possibility of         *
+* allowing symbols that weren't already covered by 'other'                    *
+*                                                                             *
+* Assumes that other->mod == this->mod.  Will not verify that fact.           *
+*                                                                             *
+************************************** | *************************************/
+
+bool ImportStmt::providesNewSymbols(const UseStmt* other) const {
+  if (other->isPlainUse()) {
+    // Other is a general use, without an 'only' or 'except' list.  It covers
+    // everything we could possibly cover, so we don't provide new symbols.
+    return false;
+  }
+
+  if (!providesQualifiedAccess()) {
+    // UseStmts always provide qualified access, and we don't provide any
+    // symbols for unqualified access, so by definition we are covered!
+    return false;
+  }
+
+  // Otherwise, we provide symbols for unqualified access, so we might provide
+  // something that the limited UseStmt doesn't.
+  if (other->hasExceptList()) {
+    // If there's overlap between our symbols and the other's except list, then
+    // we provide new symbols
+    unsigned int numSame = 0;
+    for_vector(const char, name, unqualified) {
+      if (std::find(other->named.begin(), other->named.end(),
+                    name) != other->named.end()) {
+        numSame++;
+      }
+    }
+    return numSame > 0;
+
+  } else if (other->named.size() + other->renamed.size() < unqualified.size()) {
+    // Other has an 'only' list and it has less symbols in it than our list of
+    // unqualified symbols.  By definition, this means we are providing symbols
+    // not available in other.
+    return true;
+
+  } else {
+    unsigned int numSame = 0;
+
+    for_vector(const char, name, unqualified) {
+      if (std::find(other->named.begin(), other->named.end(),
+                    name) != other->named.end()) {
+        numSame++;
+      }
+    }
+
+    // If all of our symbols for unqualified access were in the other's 'only'
+    // list, then we don't provide anything new.
+    return numSame != unqualified.size();
```

Comment:
Do you check somewhere for things like `use M only x, x` ? If not, this code is not correct.

Reply:
I do check, but it looks like I only warn on that case (see test/visibility/[only|except]/repeatInList.chpl).  However, I believe this code will stop after the first time it finds a match, so a use statement having multiple with the same name will only mean we do extra work to check if we match everything provided